core/store/src/flat_state.rs

```diff
@@ -9,34 +9,134 @@
     /// because they are the same for each shard and they are requested only
     /// once during applying chunk.
-    // TODO (#7327): implement flat state deltas to support forks.
-    // TODO (#7327): store on top of final head (or earlier) so updates will
-    // only go forward.
+    // TODO (#7327): lock flat state when `get_ref` is called or tail is being updated. Otherwise, `apply_chunks` and
+    // `postprocess_block` parallel execution may corrupt the state.
     #[derive(Clone)]
     pub struct FlatState {
+        /// Used to access flat state stored at the head of flat storage.
+        /// It should store all trie keys and values/value refs for the state on top of
+        /// flat_storage_state.head, except for delayed receipt keys.
         store: Store,
+        /// Id of the shard which state is accessed by this object.
+        shard_uid: ShardUId,
+        /// The block for which key-value pairs of its state. Note that it is not necessarily
+        /// the latest block.
+        head: CryptoHash,
     }
 
     impl FlatState {
-        fn get_raw_ref(&self, key: &[u8]) -> Result<Option<crate::db::DBSlice<'_>>, StorageError> {
-            return self
+        /// Possibly creates a new [`FlatState`] object backed by given storage.
+        ///
+        /// Always returns `None` if the `protocol_feature_flat_state` Cargo feature is
+        /// not enabled.  Otherwise, returns a new [`FlatState`] object backed by
+        /// specified storage.
+        pub fn maybe_new(
+            shard_uid: ShardUId,
+            prev_block_hash: &CryptoHash,
+            store: &Store,
+        ) -> Option<FlatState> {
+            Some(FlatState { store: store.clone(), shard_uid, head: prev_block_hash.clone() })
+        }
+
+        /// Update the tail of the flat storage. Return a StoreUpdate for the disk update.
+        pub fn update_tail(
+            shard_uid: ShardUId,
+            block_hash: &CryptoHash,
+            store: &Store,
+        ) -> StoreUpdate {
+            let mut store_update = StoreUpdate::new(store.storage.clone());
+            store_update
+                .set_ser(DBCol::FlatStateMisc, &shard_uid.try_to_vec().unwrap(), block_hash)
+                .expect("Borsh cannot fail");
+            store_update
+        }
+
+        /// Get deltas for blocks between flat state tail and current head.
+        /// If sequence of deltas contains final block, tail is moved to this tail and all deltas until the tail are
+        /// applied.
+        fn get_deltas_between_blocks(&self) -> Result<Vec<FlatStateDelta>, StorageError> {
+            let target_block_hash = self.head;
+            let flat_state_tail: CryptoHash = self
                 .store
-                .get(crate::DBCol::FlatState, key)
-                .map_err(|_| StorageError::StorageInternalError);
+                .get_ser(DBCol::FlatStateMisc, &self.shard_uid.try_to_vec().unwrap())
```

Comment:
Why do you need to get the deltas between tail and head? We need the deltas between head and the current block that get_ref is requesting. `head` is the block where where FlatState stores the key value state at, and `tail` is the last block that FlatState will keep FlatStateDelta. If we use the last final head as head, then `tail` is only needed to gc FlatStateDelta, and we don't need it now.

Reply:
I see, let's do the rename it and also add comments to that field explaining that all `get_ref`s are called on the same block.